### DISCUSSION TITLE
Pico W and amount of udp sockets

### DISCUSSION THREAD

**Discussion by manpowre** [micropython/micropython#11105] [RP2040 / Pico]
On Pico W, I have code that opens a few udp sockets for different messages:
udp_in_socket
udp_debug_out_socket
udp_step_out_socket
udp_UI_out_socket

Unfortunately, opening these 4 sockets fails with micropython and Pico W.
Error message:
OSError: [Errno 12] ENOMEM

I tried the latest nightly build as of today. On Arduino Nano rp2040 this works just fine.

Offcourse I removed the debug socket to get all of my code to work, but its not optimal offcourse. I can always reuse the other sockets to send different types of messages, but I wanted to report this if someone else runs into this.

Anyone know why, or is this a bug ?

Here is code that reproduces the issue:

```py
import network
from socket import *
import micropython

import gc
import os

def df():
  s = os.statvfs('//')
  return ('{0} MB'.format((s[0]*s[3])/1048576))

def free(full=False):
  F = gc.mem_free()
  A = gc.mem_alloc()
  T = F+A
  P = '{0:.2f}%'.format(F/T*100)
  if not full: return P
  else : return ('Total:{0} Free:{1} ({2})'.format(T,F,P))

IP_ADDRESS = "192.168.1.30" 
UDP_ADDRESS = "224.0.0.1"

nic = None 
nic = network.WLAN(network.AP_IF)

nic.active(False)
print(dir(network))
print(dir(network.WLAN))

nic.config(essid="wifi")
nic.config(password="pass")
nic.config(channel=5)

nic.active(True)

print("essid", nic.config("essid"))
print("ssid", nic.config("ssid"))
print("channel", nic.config("channel"))

print(free())

udp_in_socket = socket(AF_INET, SOCK_DGRAM) 
udp_in_socket.settimeout(0.05)
udp_debug_out_socket = socket(AF_INET, SOCK_DGRAM)
udp_debug_out_socket.settimeout(0.2)
udp_step_out_socket = socket(AF_INET, SOCK_DGRAM)
udp_step_out_socket.settimeout(0.2)
print(free())
udp_UI_out_socket = socket(AF_INET, SOCK_DGRAM)
udp_UI_out_socket.settimeout(2)

udp_in_address = getaddrinfo(IP_ADDRESS, 5501)[0][-1] 
debug_info_out_address = (UDP_ADDRESS, 5503)
step_udp_out_address = (UDP_ADDRESS, 5502) 

udp_in_socket.bind(udp_in_address)

```

(Edit by @jimmo to fix formatting)

**Comment by smudge1977** [comment]
`gc.collect()` may well be your friend - I found with that same free function even an event loop doing nothing was eating away at memory....

**Comment by jimmo** [comment]
@manpowre There's a limited amount of RAM available to LWIP. See `ports/rp2/lwip_inc/lwipopts.h` -- I think you need to increase `MEMP_NUM_UDP_PCB`

(Also please encode code in triple-backticks as per the Welcome post)
